crates/nostr/src/nips/nip26.rs

```diff
@@ -43,73 +107,441 @@ pub fn sign_delegation(
+        match self.to_json() {
+            Err(e) => write!(f, "(error {e})"),
+            Ok(s) => write!(f, "{s}"),
+        }
+    }
+}
+
+impl FromStr for DelegationTag {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Self::from_json(s)
+    }
+}
+
+/// A condition from the delegation conditions.
+#[derive(Clone)]
+pub(crate) enum Condition {
+    /// Event kind, e.g. kind=1
+    Kind(u64),
+    /// Creation time before, e.g. created_at<1679000000
+    CreatedBefore(u64),
+    /// Creation time after, e.g. created_at>1676000000
+    CreatedAfter(u64),
+}
+
+/// Set of conditions of a delegation.
+#[derive(Clone)]
+pub(crate) struct Conditions {
+    cond: Vec<Condition>,
+}
+
+/// Represents properties of an event, relevant for delegation
+pub struct EventProperties {
+    /// Event kind. For simplicity/flexibility, numeric type is used.
+    kind: u64,
+    /// Creation time, as unix timestamp
+    created_time: u64,
+}
+
+impl Condition {
+    /// Evaluate whether an event satisfies this condition
+    pub(crate) fn evaluate(&self, ep: &EventProperties) -> Result<(), ValidationError> {
+        match self {
+            Self::Kind(k) => {
+                if ep.kind != *k {
+                    return Err(ValidationError::InvalidKind);
+                }
+            }
+            Self::CreatedBefore(t) => {
+                if ep.created_time >= *t {
+                    return Err(ValidationError::CreatedTooLate);
+                }
+            }
+            Self::CreatedAfter(t) => {
+                if ep.created_time <= *t {
+                    return Err(ValidationError::CreatedTooEarly);
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl ToString for Condition {
+    fn to_string(&self) -> String {
+        match self {
+            Self::Kind(k) => format!("kind={k}"),
+            Self::CreatedBefore(t) => format!("created_at<{t}"),
+            Self::CreatedAfter(t) => format!("created_at>{t}"),
+        }
+    }
+}
+
+impl FromStr for Condition {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let try_kind = s.strip_prefix("kind=");
+        if try_kind.is_some() {
+            let n = u64::from_str(try_kind.unwrap())?;
+            return Ok(Condition::Kind(n));
+        }
+        let try_created_before = s.strip_prefix("created_at<");
+        if try_created_before.is_some() {
+            let n = u64::from_str(try_created_before.unwrap())?;
+            return Ok(Condition::CreatedBefore(n));
+        }
+        let try_created_after = s.strip_prefix("created_at>");
+        if try_created_after.is_some() {
+            let n = u64::from_str(try_created_after.unwrap())?;
+            return Ok(Condition::CreatedAfter(n));
+        }
+        Err(Error::ConditionsParseInvalidCondition)
+    }
+}
+
+impl Conditions {
+    pub fn new() -> Self {
+        Conditions { cond: Vec::new() }
```

Comment:
Use `Self` instead of `Conditions` (I flagged only this line but apply this change also in the other methods) :)